applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h

```diff
@@ -114,44 +113,201 @@ class MapperVertexMorphing : public Mapper
     ///@{
 
     // --------------------------------------------------------------------------
-    void InitializeMapping() override
+    void Initialize() override
     {
-        CreateListOfNodesOfDesignSurface();
-        CreateFilterFunction();
-        InitializeMappingVariables();
-        AssignMappingIds();
+        BuiltinTimer timer;
+        std::cout << "> Starting initialization of mapping..." << std::endl;
+
+        if (mIsMappingInitialized == false)
+        {
+            CreateListOfNodesInOriginMdpa();
+            CreateFilterFunction();
+            InitializeMappingVariables();
+            AssignMappingIds();
+        }
+
+        InitializeComputationOfMappingMatrix();
+        CreateSearchTreeWithAllNodesInOriginMdpa();
+        ComputeMappingMatrix();
+
+        mIsMappingInitialized = true;
```

Comment:
Therefore you just call again the initialize function.

Reply:
Therefore you just call again the initialize function.

Reply:
Hm but Initialize is only called at the beginning of the optimization loop. Does that mean we are not recalculating the mapping matrix in each step?

Reply:
Ok I see now that you changed the python files where the mapper is called. This is confusing, a few points i don't like a lot about this design:
- The initialize function should always do the same and not internally have a check.
- To do so, one could split it into two functions:
``` cpp
Initialize() //everything that needs to be done once
Reset/Update() //everything that needs to be done in each step
```

- The fact that in the python files, mapper.Initialize() is called in one of the mapping functions is also not clean.

- Please add a Finalize function, even if it may be empty (I was thought in one of my recent PRs that if there is an initialize there should also be a finalize, for the sake of completeness).

Reply:
The first three comments, I totally agree and I changed the code accordingly. Thanks a lot for that careful review!

However, In my opinion, there is no need for a finalize function. Let me try to justify this a bit. All written code should be there for a reason. Otherwise, it is literally useless and will just blow up the overall code with all consequences, which is a bad programming style. In case of the mapper, I agree, that as soon as one derived mapper needs a finalization, it should be there and included in the base class (and also be called in the python algorithms). But, as long as none of them are requiring it, it will be code for nothing, which is even worse than not being consistent. Anyways, consistent with what? Random other classes, that have nothing to do with the mapper, but do have an Initialize and Finalize?

So I think its fine to just leave it for now. @armingeiser What do you think?

Reply:
I see your points (was also my opinion before Mike convinced me). I don't consider this point important, so I'm fine continuing without Finalize.